Switch to TypeScript parser

karlhorky opened the pull request:
Closes #3

karlhorky committed:
Switch to TypeScript parser

Fixes #3

Lisba commented:
Hi @karlhorky I really appreciate your interest for improving this plugin. I would be really glad to accept your proposed change but I cannot see how this fix the issue you opened #3 .

I uploaded a video during the test, showing that this doesn't fix the mentioned issue. If you think my test is wrong, could you please upload any test that confirms this as the right fix for the mentioned issue?

https://github.com/user-attachments/assets/fdab8483-22f4-4b5a-83b5-9b139a80d97b

Lisba reviewed (changes requested):
I tried this changes but I didn't see the issue fixed. Could you please add a test that confirms this as the right fix for the mentioned issue?

karlhorky commented:
@Lisba maybe you modified the wrong file in `node_modules`?

Here's my [`patch-package`](https://www.npmjs.com/package/patch-package) patch, it changes `node_modules/prettier-plugin-unused-imports-configurable/dist/index.js` (in the `dist` folder):

`patches/prettier-plugin-unused-imports-configurable+1.14.3.patch`

```patch
diff --git a/node_modules/prettier-plugin-unused-imports-configurable/dist/index.js b/node_modules/prettier-plugin-unused-imports-configurable/dist/index.js
index ab10376..9349a03 100644
--- a/node_modules/prettier-plugin-unused-imports-configurable/dist/index.js
+++ b/node_modules/prettier-plugin-unused-imports-configurable/dist/index.js
@@ -1 +1,9 @@
-"use strict";var e,t,r,n,o,i,s,a,c,l,p=require("prettier/parser-babel"),u=require("ts-morph");function f(e){return e&&e.__esModule&&Object.prototype.hasOwnProperty.call(e,"default")?e.default:e}function g(){if(t)return e;t=1;const r=(e,t,r,n)=>{if(!r)throw new Error("Module specifier is required");if(!e&&t.length<1&&!n)throw new Error("No imports received")},n=e=>e.startsWith("'")||e.startsWith('"')?e:`'${e}'`,o=(e,t,r)=>{const n=e?`${e}`:"",o=t?.length>0?`{ ${t.join(", ")} }`:"",i=r?`* as ${r}`:"";return{defaultPart:n,namedPart:o,namespacePart:i,separator:n&&(o||i)?", ":""}};return e=(e,t,i,s=null)=>{r(e,t,i,s);const a=n(i),{defaultPart:c,namedPart:l,namespacePart:p,separator:u}=o(e,t,s);return`import ${c}${u}${l||p} from ${a};`}}function d(){if(a)return s;a=1;const{Project:e}=u,t=g(),c=function(){if(n)return r;n=1;const{ts:e}=u,t=e=>{const t=e.getImportDeclarations();return 0===t.length?{start:0,end:0}:{start:t[0].getStart(),end:t[t.length-1].getEnd()}},o=(t,r)=>t.getDescendants().filter((t=>{if(t.getKind()!==e.SyntaxKind.Identifier||t.getText()!==r)return!1;const n=t.getParent();return n?.getKind()!==e.SyntaxKind.PropertyAccessExpression||n.getName()!==r})),i=(e,t)=>e.filter((e=>{const r=e.getStart();return r<t.start||r>=t.end})).length;return r=(e,r)=>{const n=o(e,r),s=t(e);return i(n,s)}}(),l=function(){if(i)return o;i=1;const e=g();function t(t,r){let n=null;const o=new Set;for(const e of t){const t=e.getDefaultImport()?.getText();t&&(n=t);for(const t of e.getNamedImports())o.add(t.getText())}t[0].replaceWithText(e(n,[...o],r));for(let e=1;e<t.length;e++)t[e].remove()}return o=e=>{const r=e.getImportDeclarations(),n=new Map;for(const e of r){const t=e.getModuleSpecifier().getText();n.has(t)||n.set(t,[]),n.get(t).push(e)}for(const[e,r]of n.entries()){const n=[];for(const e of r){const t=e.getNamespaceImport()?.getText();t||n.push(e)}n.length>1&&t(n,e)}}}(),p=(e,t)=>{if(!t?.filepath)return console.error("Filepath is undefined. Skipping..."),!1;if(!e||""===e.trim())return console.error("File content is empty. Skipping..."),!1;return(Array.isArray(t?.ignoreDirectories)?t?.ignoreDirectories:[]).some((e=>t.filepath.includes(e)))?(console.error(`Skipping file inside an ignored directory: ${t.filepath}`),!1):!f(e)||(console.error("Exclusion comment found. Skipping file..."),!1)},f=e=>{const t=/^\s*\/\/\s*prettier-ignore-unused-imports-configurable\s*$/,r=e.split("\n");let n=!1;for(let e=0;e<r.length;e++){const o=r[e].trim();if(o.startsWith("import"))return n;t.test(o)&&(n=!0)}return!1},d=e=>{e.getImportDeclarations().forEach((r=>{let n=!1;const o=r.getNamespaceImport()?.getText();let i=r.getDefaultImport()?.getText();const s=r.getNamedImports().map((e=>({name:e.getName(),text:e.getText(),alias:e.getAliasNode()?.getText()}))),a=r.getModuleSpecifier().getText();i&&c(e,i)<1&&(i=null,n=!0);const l=s.filter((t=>c(e,t.alias||t.name)<1)),p=s.filter((e=>!l.includes(e)));l.length>0&&(n=!0),o&&c(e,o)<1?r.remove():n&&(0!==p.length||i?r.replaceWithText(t(i,p.map((e=>e.text)),a)):r.remove())}))};return s=(t,r)=>{if(!p(t,r))return t;try{const n=(new e).createSourceFile(r.filepath,t,{overwrite:!0});return l(n),d(n),n.getFullText()}catch(e){return console.error("Error processing file:",r.filepath,e),t}}}var m=f(function(){if(l)return c;l=1;const{parsers:e}=p,t=d();return c={parsers:{typescript:{...e["babel-ts"],preprocess:t},javascript:{...e.babel,preprocess:t}},options:{ignoreDirectories:{type:"string",array:!0,default:[{value:[]}],category:"Global",description:"Directories to ignore for unused import cleaning."}}}}());module.exports=m;
+"use strict";var e,t,r,n,o,i,s,a,c,l,
+
+typescriptParsers=require("prettier/parser-typescript").parsers,
+
+p=require("prettier/parser-babel"),u=require("ts-morph");function f(e){return e&&e.__esModule&&Object.prototype.hasOwnProperty.call(e,"default")?e.default:e}function g(){if(t)return e;t=1;const r=(e,t,r,n)=>{if(!r)throw new Error("Module specifier is required");if(!e&&t.length<1&&!n)throw new Error("No imports received")},n=e=>e.startsWith("'")||e.startsWith('"')?e:`'${e}'`,o=(e,t,r)=>{const n=e?`${e}`:"",o=t?.length>0?`{ ${t.join(", ")} }`:"",i=r?`* as ${r}`:"";return{defaultPart:n,namedPart:o,namespacePart:i,separator:n&&(o||i)?", ":""}};return e=(e,t,i,s=null)=>{r(e,t,i,s);const a=n(i),{defaultPart:c,namedPart:l,namespacePart:p,separator:u}=o(e,t,s);return`import ${c}${u}${l||p} from ${a};`}}function d(){if(a)return s;a=1;const{Project:e}=u,t=g(),c=function(){if(n)return r;n=1;const{ts:e}=u,t=e=>{const t=e.getImportDeclarations();return 0===t.length?{start:0,end:0}:{start:t[0].getStart(),end:t[t.length-1].getEnd()}},o=(t,r)=>t.getDescendants().filter((t=>{if(t.getKind()!==e.SyntaxKind.Identifier||t.getText()!==r)return!1;const n=t.getParent();return n?.getKind()!==e.SyntaxKind.PropertyAccessExpression||n.getName()!==r})),i=(e,t)=>e.filter((e=>{const r=e.getStart();return r<t.start||r>=t.end})).length;return r=(e,r)=>{const n=o(e,r),s=t(e);return i(n,s)}}(),l=function(){if(i)return o;i=1;const e=g();function t(t,r){let n=null;const o=new Set;for(const e of t){const t=e.getDefaultImport()?.getText();t&&(n=t);for(const t of e.getNamedImports())o.add(t.getText())}t[0].replaceWithText(e(n,[...o],r));for(let e=1;e<t.length;e++)t[e].remove()}return o=e=>{const r=e.getImportDeclarations(),n=new Map;for(const e of r){const t=e.getModuleSpecifier().getText();n.has(t)||n.set(t,[]),n.get(t).push(e)}for(const[e,r]of n.entries()){const n=[];for(const e of r){const t=e.getNamespaceImport()?.getText();t||n.push(e)}n.length>1&&t(n,e)}}}(),p=(e,t)=>{if(!t?.filepath)return console.error("Filepath is undefined. Skipping..."),!1;if(!e||""===e.trim())return console.error("File content is empty. Skipping..."),!1;return(Array.isArray(t?.ignoreDirectories)?t?.ignoreDirectories:[]).some((e=>t.filepath.includes(e)))?(console.error(`Skipping file inside an ignored directory: ${t.filepath}`),!1):!f(e)||(console.error("Exclusion comment found. Skipping file..."),!1)},f=e=>{const t=/^\s*\/\/\s*prettier-ignore-unused-imports-configurable\s*$/,r=e.split("\n");let n=!1;for(let e=0;e<r.length;e++){const o=r[e].trim();if(o.startsWith("import"))return n;t.test(o)&&(n=!0)}return!1},d=e=>{e.getImportDeclarations().forEach((r=>{let n=!1;const o=r.getNamespaceImport()?.getText();let i=r.getDefaultImport()?.getText();const s=r.getNamedImports().map((e=>({name:e.getName(),text:e.getText(),alias:e.getAliasNode()?.getText()}))),a=r.getModuleSpecifier().getText();i&&c(e,i)<1&&(i=null,n=!0);const l=s.filter((t=>c(e,t.alias||t.name)<1)),p=s.filter((e=>!l.includes(e)));l.length>0&&(n=!0),o&&c(e,o)<1?r.remove():n&&(0!==p.length||i?r.replaceWithText(t(i,p.map((e=>e.text)),a)):r.remove())}))};return s=(t,r)=>{if(!p(t,r))return t;try{const n=(new e).createSourceFile(r.filepath,t,{overwrite:!0});return l(n),d(n),n.getFullText()}catch(e){return console.error("Error processing file:",r.filepath,e),t}}}var m=f(function(){if(l)return c;l=1;const{parsers:e}=p,t=d();return c={parsers:{
+
+  typescript:{...typescriptParsers.typescript,preprocess:t},
+
+javascript:{...e.babel,preprocess:t}},options:{ignoreDirectories:{type:"string",array:!0,default:[{value:[]}],category:"Global",description:"Directories to ignore for unused import cleaning."}}}}());module.exports=m;
```

Here's a video of Prettier formatting to a single line:

https://github.com/user-attachments/assets/579e1206-ba03-4afd-896f-1f0b2ea787fe

Lisba commented:
@karlhorky Thank you, what happened was not exactly what you described but it helped me to find the real problem. In development I was using `pnpm link` to test the plugin on a "consumer" project, the problem was that it was using the builded version of the package, so activating the rollup's watch mode let me see the plugin changes in realtime.

The PR is approved and will be merge. Thank you for investing time to highlight this issue and find a solution, too. I really appreciate it! 🫶
